### PR TITLE
tracing: fix the fast-path in the presence of an otel collector, and add tests

### PR DESCRIPTION
**Pull request by guswynn**
The fast path has been broken in production for a LONG time, but a bug in tracing (tracked internally here https://github.com/MaterializeInc/materialize/issues/16556) meant that the code as laid out by https://github.com/MaterializeInc/materialize/pull/16443 worked _in ci_. This pr works around the regression that erroneously-only-somtimes-exists by correctly filtering the sentry-layer. Note I also add an endpoint to check the current fast-path, and also revert https://github.com/MaterializeInc/materialize/pull/16443, as its no longer broken. I feel comfortable with this because of the added tests.

The behavior I noticed that inspired this pr was that the sentry layer would be starved of events more verbose than `DEBUG`/`TRACE`, but only if the otel layer was `None`. We never noticed because it only cares about `INFO` and above. 

In practice, when @jpepin enables the otel layer in production, we will see a perf hit as it will eliminate the fast path this pr adds back.

Also, i added tests to ensure our ci and prod situation dont regress. I chose an mzcompose because, and I think @benesch will agree with this, i dont trust any test setup that doesnt literally mirror production exactly.

### Motivation

  * This PR fixes a previously unreported bug.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

### CODE REVIEW

File: src/http-util/src/lib.rs [hunk 1]
```diff
@@ -128,3 +129,14 @@ pub async fn handle_modify_filter_target(
         Err(e) => (StatusCode::BAD_REQUEST, e.to_string()),
     }
 }
+
+/// Allows dynamic control of the stderr log filter
+#[allow(clippy::unused_async)]
+pub async fn generate_tracing_info() -> impl IntoResponse {
```
Comment (benesch) [inline src/http-util/src/lib.rs:135]: For consistency with the other handlers:

```suggestion
pub async fn handle_tracing() -> impl IntoResponse {
```

File: src/http-util/src/lib.rs [hunk 2]
```diff
@@ -128,3 +129,14 @@ pub async fn handle_modify_filter_target(
...
+    (
+        StatusCode::OK,
+        Json(json!({
+            "filter_fast_path": tracing::level_filters::LevelFilter::current().to_string()
```
Comment (benesch) [inline src/http-util/src/lib.rs:139]: Should this be called `current_level_filter` to match the Rust API?

```suggestion
            "current_level_filter": tracing::level_filters::LevelFilter::current().to_string()
```

File: src/http-util/src/lib.rs [hunk 3]
```diff
@@ -128,3 +129,14 @@ pub async fn handle_modify_filter_target(
         Err(e) => (StatusCode::BAD_REQUEST, e.to_string()),
     }
 }
+
+/// Allows dynamic control of the stderr log filter
```
Comment (benesch) [inline src/http-util/src/lib.rs:133]: Copypasta

File: src/ore/src/tracing/mod.rs [hunk 4]
```diff
@@ -368,7 +369,35 @@ where
...
     let stack = stack.with(tokio_console_layer);
-    let stack = stack.with(sentry_tracing::layer().event_filter(sentry_event_filter));
+    let stack = stack.with(
+        sentry_tracing::layer()
+            .event_filter(sentry_event_filter)
+            // WARNING, ENTERING THE SPOOKY ZONE
+            //
+            // While sentry provides an event filter above that maps events to types of sentry events, its `Layer`
+            // implementation does not participate in `tracing`'s level-fast-path implementation, which depends on
+            // a hidden api (<https://docs.rs/tracing-subscriber/latest/src/tracing_subscriber/layer/mod.rs.html#864-867>)
```
Comment (benesch) [inline src/ore/src/tracing/mod.rs:379]: Can you link to a permalink on GitHub, so this doesn't go stale the next time a new release of `tracing_subscriber` is released?

File: src/ore/src/tracing/mod.rs [hunk 5]
```diff
@@ -368,7 +369,35 @@ where
...
+            // which is primarily manged by filters (like below). The fast path skips verbose log
+            // (and span) levels that no layer is interested by reading a single atomic. Usually, not implementing this
+            // api means "give me everything, including `trace`, unless you attach a filter to me.
+            //
+            // The curious thing here (and a bug in tracing) is that _some configurations of our layer stack above_,
+            // if you don't have this filter can cause the fast-path to trigger, despite the fact
```
Comment (benesch) [inline src/ore/src/tracing/mod.rs:385]: Ah, so the fact that the fast path was _ever_ enabled was the bug? Wild.

File: test/tracing/mzcompose.py [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+            "-H",
+            "Content-Type: application/json",
+            "0:6878/api/stderr/config",
+            "-d",
+            '{"targets": "info"}',
+        )
+        value = json.loads(
+            c.exec("materialized", "curl", "0:6878/api/tracing", capture=True).stdout
+        )
+        assert value["filter_fast_path"] == "info"
```
Comment (benesch) [inline test/tracing/mzcompose.py:150]: Can you pivot this to pull out the internal port via `c.port` and then hit that port with the `requests` Python library? It'll be much easier to read than the `curl` invocations!

**Review by benesch — COMMENTED** [review]
Thank you so much for tracking this one down, @guswynn!

File: src/ore/src/tracing/mod.rs (hunk 5 above)
Comment (guswynn) [reply src/ore/src/tracing/mod.rs:385]: YES

File: test/tracing/mzcompose.py [hunk 7]
```diff
@@ -0,0 +1,100 @@
+# Copyright Materialize, Inc. and contributors. All rights reserved.
+#
+# Use of this software is governed by the Business Source License
+# included in the LICENSE file at the root of this repository.
+#
+# As of the Change Date specified in that file, in accordance with
+# the Business Source License, use of this software will be governed
+# by the Apache License, Version 2.0.
+
+# Tests the dynamic tracing setup on environmentd
```
Comment (benesch) [inline test/tracing/mzcompose.py:10]: Can you make this a documentation comment in the module instead? Then it should show up in `./mzcompose list` or whatever the command is.

**Review by benesch — APPROVED** [review]
Nice! That test is really clean now. Thank you, Gus!

File: test/tracing/mzcompose.py (hunk 7 above)
Comment (guswynn) [reply test/tracing/mzcompose.py:10]: oh lmao i forgot about that in python lol

Comment (guswynn) [reply test/tracing/mzcompose.py:10]: I think only function-level ones show up so ill add it to `workflow_default`

**Pull request merged**